File: sdk/python/lavs_runtime/permission_checker.py
"""
LAVS Permission Checker.

Enforces permission constraints declared in LAVS manifests.

Enforcement Model:
- Path traversal: ENFORCED - handler cwd and command paths validated
- fileAccess: ADVISORY - glob pattern matching at dispatch time
- networkAccess: ADVISORY - declared for auditing
- maxExecutionTime: ENFORCED - handler killed on timeout
- maxMemory: ADVISORY - not enforced in current runtime
"""

from __future__ import annotations

from pathlib import Path

from pathspec import PathSpec

from lavs_types import LAVSError, LAVSErrorCode, Permissions, ScriptHandler


class PermissionChecker:
    """
    LAVS Permission Checker - enforces manifest-declared security constraints.
    """

    def merge_permissions(
        self,
        manifest_permissions: Permissions | None,
        endpoint_permissions: Permissions | None,
    ) -> Permissions:
        """
        Merge manifest-level and endpoint-level permissions.

        Endpoint-level permissions take precedence over manifest-level.

        Args:
            manifest_permissions: Manifest-level permissions (defaults).
            endpoint_permissions: Endpoint-level permissions (overrides).

        Returns:
            Merged permissions object.
        """
        if not manifest_permissions and not endpoint_permissions:
            return Permissions()

        mp = manifest_permissions or Permissions()
        ep = endpoint_permissions or Permissions()

        return Permissions(
            file_access=ep.file_access if ep.file_access is not None else mp.file_access,
            network_access=(
                ep.network_access if ep.network_access is not None else mp.network_access
            ),
            max_execution_time=(
                ep.max_execution_time
                if ep.max_execution_time is not None
                else mp.max_execution_time
            ),
            max_memory=(
                ep.max_memory if ep.max_memory is not None else mp.max_memory
            ),
        )

    def check_path_traversal(self, target_path: str, allowed_base: str) -> None:
        """
        Check if target path is within allowed base directory.

        Prevents path traversal attacks (e.g., ../../etc/passwd).

        Args:
            target_path: The path to check (absolute or relative).
            allowed_base: The allowed base directory (absolute).

        Raises:
            LAVSError: With code PermissionDenied if path is outside allowed base.
        """
        resolved_target = (Path(allowed_base).resolve() / target_path).resolve()
        normalized_base = Path(allowed_base).resolve()

        try:
            resolved_target.relative_to(normalized_base)
        except ValueError:
            raise LAVSError(
                LAVSErrorCode.PermissionDenied,
                f"Path traversal detected: '{target_path}' resolves outside allowed directory '{allowed_base}'",
                {"resolved_path": str(resolved_target), "allowed_base": str(normalized_base)},
            )

    def check_handler_cwd(self, handler: ScriptHandler, agent_dir: str) -> None:
        """
        Check if handler's working directory is within allowed base.

        Args:
            handler: Script handler with optional cwd.
            agent_dir: Agent directory (the allowed base).

        Raises:
            LAVSError: If cwd is outside agent directory.
        """
        if not handler.cwd:
            return

        self.check_path_traversal(handler.cwd, agent_dir)

    def check_file_access(self, file_path: str, permissions: Permissions) -> bool:
        """
        Check if file access patterns allow a given path.

        Uses glob matching against permissions.file_access patterns.

        Args:
            file_path: The file path to check (relative to agent dir).
            permissions: Permission constraints including file_access patterns.

        Returns:
            True if access is allowed, False otherwise.
        """
        if not permissions.file_access or len(permissions.file_access) == 0:
            return True

        # PathSpec gitignore expects paths without leading ./
        match_path = file_path.lstrip("./") if file_path.startswith("./") else file_path

        # Check deny patterns first (negative patterns take precedence)
        for pattern in permissions.file_access:
            if pattern.startswith("!"):
                pat = pattern[1:].lstrip("./")
                spec = PathSpec.from_lines("gitignore", [pat])
                if spec.match_file(match_path):
                    return False

        # Check allow patterns
        for pattern in permissions.file_access:
            if not pattern.startswith("!"):
                pat = pattern.lstrip("./")
                spec = PathSpec.from_lines("gitignore", [pat])
                if spec.match_file(match_path):
                    return True

        return False

    def get_effective_timeout(
        self,
        handler: ScriptHandler,
        permissions: Permissions,
        default_timeout: int = 30000,
    ) -> int:
        """
        Get effective execution timeout for a handler.

        Priority: handler.timeout > permissions.max_execution_time > default.

        Args:
            handler: Script handler with optional timeout.
            permissions: Permission constraints.
            default_timeout: Default timeout in milliseconds.

        Returns:
            Effective timeout in milliseconds.
        """
        if handler.timeout is not None and handler.timeout > 0:
            if (
                permissions.max_execution_time is not None
                and permissions.max_execution_time > 0
            ):
                return min(handler.timeout, permissions.max_execution_time)
            return handler.timeout

        if (
            permissions.max_execution_time is not None
            and permissions.max_execution_time > 0
        ):
            return permissions.max_execution_time

        return default_timeout

    def assert_allowed(
        self,
        handler: ScriptHandler,
        permissions: Permissions,
        agent_dir: str,
    ) -> None:
        """
        Run all permission checks for script handler execution.

        Args:
            handler: Script handler configuration.
            permissions: Merged permissions.
            agent_dir: Agent base directory.

        Raises:
            LAVSError: If any permission check fails.
        """
        self.check_handler_cwd(handler, agent_dir)

        if handler.command and ("/" in handler.command or "\\" in handler.command):
            if not Path(handler.command).is_absolute():
                self.check_path_traversal(handler.command, agent_dir)

File: sdk/typescript/runtime/src/permission-checker.ts
/**
 * LAVS Permission Checker
 *
 * Enforces permission constraints declared in LAVS manifests.
 *
 * ## Enforcement Model
 *
 * LAVS permissions have two enforcement levels:
 *
 * | Permission       | Level    | What is checked                                |
 * |-----------------|----------|------------------------------------------------|
 * | Path traversal  | ENFORCED | Handler cwd and command paths are validated     |
 * | fileAccess      | ADVISORY | Glob pattern matching at dispatch time;         |
 * |                 |          | NOT intercepted at OS/syscall level during      |
 * |                 |          | script execution. Scripts can still access       |
 * |                 |          | files the process has OS-level permission for.   |
 * | networkAccess   | ADVISORY | Declared for auditing; not enforced at runtime. |
 * |                 |          | Use OS-level isolation (nsjail, Docker) for     |
 * |                 |          | strict enforcement.                              |
 * | maxExecTime     | ENFORCED | Handler killed via SIGTERM/SIGKILL on timeout.  |
 * | maxMemory       | ADVISORY | Not enforced in current runtime.                |
 *
 * Future: Consider OS-level sandboxing for fileAccess and networkAccess.
 */

import path from 'path';
import { minimatch } from 'minimatch';
import {
  Permissions,
  ScriptHandler,
  LAVSError,
  LAVSErrorCode,
} from './types';

/**
 * LAVS Permission Checker - enforces manifest-declared security constraints
 */
export class PermissionChecker {
  /**
   * Merge manifest-level and endpoint-level permissions.
   * Endpoint-level permissions take precedence over manifest-level.
   *
   * @param manifestPermissions - Manifest-level permissions (defaults)
   * @param endpointPermissions - Endpoint-level permissions (overrides)
   * @returns Merged permissions object
   */
  mergePermissions(
    manifestPermissions?: Permissions,
    endpointPermissions?: Permissions
  ): Permissions {
    if (!manifestPermissions && !endpointPermissions) {
      return {};
    }
    if (!manifestPermissions) {
      return { ...endpointPermissions };
    }
    if (!endpointPermissions) {
      return { ...manifestPermissions };
    }

    return {
      fileAccess: endpointPermissions.fileAccess ?? manifestPermissions.fileAccess,
      networkAccess: endpointPermissions.networkAccess ?? manifestPermissions.networkAccess,
      maxExecutionTime: endpointPermissions.maxExecutionTime ?? manifestPermissions.maxExecutionTime,
      maxMemory: endpointPermissions.maxMemory ?? manifestPermissions.maxMemory,
    };
  }

  /**
   * Check if a target path is within the allowed base directory.
   * Prevents path traversal attacks (e.g., ../../etc/passwd).
   *
   * @param targetPath - The path to check (absolute or relative)
   * @param allowedBase - The allowed base directory (absolute)
   * @throws LAVSError with code PermissionDenied if path is outside allowed base
   */
  checkPathTraversal(targetPath: string, allowedBase: string): void {
    const resolvedTarget = path.resolve(allowedBase, targetPath);
    const normalizedBase = path.resolve(allowedBase);

    // Ensure resolved target is within or equal to the allowed base
    if (!resolvedTarget.startsWith(normalizedBase + path.sep) && resolvedTarget !== normalizedBase) {
      throw new LAVSError(
        LAVSErrorCode.PermissionDenied,
        `Path traversal detected: '${targetPath}' resolves outside allowed directory '${allowedBase}'`,
        { resolvedPath: resolvedTarget, allowedBase: normalizedBase }
      );
    }
  }

  /**
   * Check if a handler's working directory is within the allowed base.
   *
   * @param handler - Script handler with optional cwd
   * @param agentDir - Agent directory (the allowed base)
   * @throws LAVSError with code PermissionDenied if cwd is outside agent directory
   */
  checkHandlerCwd(handler: ScriptHandler, agentDir: string): void {
    if (!handler.cwd) {
      return; // No custom cwd, will use agentDir as default
    }

    this.checkPathTraversal(handler.cwd, agentDir);
  }

  /**
   * Check if file access patterns allow a given path.
   * Uses glob matching against permissions.fileAccess patterns.
   *
   * @param filePath - The file path to check (relative to agent dir)
   * @param permissions - Permission constraints including fileAccess patterns
   * @returns true if access is allowed, false otherwise
   */
  checkFileAccess(filePath: string, permissions: Permissions): boolean {
    // No fileAccess restrictions - allow all
    if (!permissions.fileAccess || permissions.fileAccess.length === 0) {
      return true;
    }

    // Normalize the file path
    const normalizedPath = filePath.startsWith('./') ? filePath : `./${filePath}`;

    // First check deny patterns (negative patterns take precedence)
    for (const pattern of permissions.fileAccess) {
      if (pattern.startsWith('!')) {
        if (minimatch(normalizedPath, pattern.slice(1))) {
          return false; // Explicitly denied
        }
      }
    }

    // Then check allow patterns
    for (const pattern of permissions.fileAccess) {
      if (!pattern.startsWith('!')) {
        if (minimatch(normalizedPath, pattern)) {
          return true; // Allowed
        }
      }
    }

    // No positive pattern matched - deny by default
    return false;
  }

  /**
   * Get the effective execution timeout for a handler.
   * Priority: handler.timeout > permissions.maxExecutionTime > default (30000ms)
   *
   * @param handler - Script handler with optional timeout
   * @param permissions - Permission constraints with optional maxExecutionTime
   * @param defaultTimeout - Default timeout in milliseconds (default: 30000)
   * @returns Effective timeout in milliseconds
   */
  getEffectiveTimeout(
    handler: ScriptHandler,
    permissions: Permissions,
    defaultTimeout: number = 30000
  ): number {
    // Handler-level timeout takes precedence
    if (handler.timeout != null && handler.timeout > 0) {
      // But cannot exceed permission-level max
      if (permissions.maxExecutionTime != null && permissions.maxExecutionTime > 0) {
        return Math.min(handler.timeout, permissions.maxExecutionTime);
      }
      return handler.timeout;
    }

    // Permission-level timeout
    if (permissions.maxExecutionTime != null && permissions.maxExecutionTime > 0) {
      return permissions.maxExecutionTime;
    }

    return defaultTimeout;
  }

  /**
   * Run all permission checks for a script handler execution.
   * This is a convenience method that combines multiple checks.
   *
   * @param handler - Script handler configuration
   * @param permissions - Merged permissions
   * @param agentDir - Agent base directory
   * @throws LAVSError if any permission check fails
   */
  assertAllowed(
    handler: ScriptHandler,
    permissions: Permissions,
    agentDir: string
  ): void {
    // 1. Check handler cwd for path traversal
    this.checkHandlerCwd(handler, agentDir);

    // 2. Check script command path if it looks like a file path
    if (handler.command && (handler.command.includes('/') || handler.command.includes('\\'))) {
      // Only check relative paths (absolute paths are resolved by loader)
      if (!path.isAbsolute(handler.command)) {
        this.checkPathTraversal(handler.command, agentDir);
      }
    }
  }
}
